src/server/replicated_world.rs

```diff
@@ -0,0 +1,172 @@
+use bevy::{
+    ecs::{
+        archetype::{ArchetypeEntity, Archetypes},
+        component::{ComponentId, ComponentTicks, Components, StorageType, Tick},
+        query::{Access, FilteredAccess},
+        storage::TableId,
+        system::{ReadOnlySystemParam, SystemMeta, SystemParam},
+        world::unsafe_world_cell::UnsafeWorldCell,
+    },
+    prelude::*,
+    ptr::Ptr,
+};
+
+use crate::core::replication::{replication_rules::ReplicationRules, Replicated};
+
+/// A [`SystemParam`] that wraps [`World`], but provides access only for replicated components.
+///
+/// We don't use [`FilteredEntityRef`](bevy::ecs::world::FilteredEntityRef) to avoid access checks
+/// and [`StorageType`] fetch (we cache this information on replicated archetypes).
+pub(crate) struct ReplicatedWorld<'w, 's> {
+    world: UnsafeWorldCell<'w>,
+    state: &'s Access<ComponentId>,
+}
+
+impl<'w> ReplicatedWorld<'w, '_> {
+    /// Extracts a component as [`Ptr`] and its ticks from a table or sparse set, depending on its storage type.
+    ///
+    /// # Safety
+    ///
+    /// The component must be present in this archetype, have the specified storage type, and be previously marked for replication.
+    pub(super) unsafe fn get_component_unchecked(
```

Comment:
Doesn't this also require that the source of the ptr doesn't get reallocated? I don't think we can make that guarantee.

Reply:
We configured the access by manually implementing `SystemParam` to prevent other systems running in parallel from writing to replicated components. And we expose only read access with the associated runtime, so it should be fine.